File: movie-matcher/backend-java/src/main/java/com/moviematcher/model/ServerMessage.java
package com.moviematcher.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import java.util.List;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes(
    {
        @JsonSubTypes.Type(
            value = ServerMessage.ParticipantJoined.class,
            name = "ParticipantJoined"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.ParticipantLeft.class,
            name = "ParticipantLeft"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.ParticipantReady.class,
            name = "ParticipantReady"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.VotingStarted.class,
            name = "VotingStarted"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.RoomLocked.class,
            name = "RoomLocked"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.NewMovie.class,
            name = "NewMovie"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.VoteRecorded.class,
            name = "VoteRecorded"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.RoundCompleted.class,
            name = "RoundCompleted"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.VotingCompleted.class,
            name = "VotingCompleted"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.NoMoreMovies.class,
            name = "NoMoreMovies"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.LikesUpdated.class,
            name = "LikesUpdated"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.MatchingEnded.class,
            name = "MatchingEnded"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.StreamingEnded.class,
            name = "StreamingEnded"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.MatchFound.class,
            name = "MatchFound"
        ),
        @JsonSubTypes.Type(
            value = ServerMessage.MatchingStarted.class,
            name = "MatchingStarted"
        ),
        @JsonSubTypes.Type(value = ServerMessage.Error.class, name = "Error"),
    }
)
public sealed interface ServerMessage {
    record ParticipantJoined(
        @JsonProperty("participant_id") String participantId
    ) implements ServerMessage {}

    record ParticipantLeft(
        @JsonProperty("participant_id") String participantId
    ) implements ServerMessage {}

    record ParticipantReady(
        @JsonProperty("participant_id") String participantId,
        @JsonProperty("ready_count") int readyCount,
        @JsonProperty("total_count") int totalCount
    ) implements ServerMessage {}

    record VotingStarted() implements ServerMessage {}

    record RoomLocked(String message) implements ServerMessage {}

    record MatchingStarted() implements ServerMessage {}

    record NewMovie(MovieData movie) implements ServerMessage {}

    record VoteRecorded(
        @JsonProperty("participant_id") String participantId,
        @JsonProperty("movie_id") String movieId,
        @JsonProperty("is_like") boolean isLike
    ) implements ServerMessage {}

    record RoundCompleted(
        @JsonProperty("round_number") int roundNumber,
        @JsonProperty("common_likes") List<String> commonLikes
    ) implements ServerMessage {}

    record VotingCompleted(
        @JsonProperty("matched_movies") List<String> matchedMovies
    ) implements ServerMessage {}

    record NoMoreMovies() implements ServerMessage {}

    record LikesUpdated(
        @JsonProperty("all_likes") List<ParticipantLikes> allLikes,
        @JsonProperty("common_likes") List<MovieData> commonLikes
    ) implements ServerMessage {}

    record MatchingEnded(
        @JsonProperty("all_likes") List<ParticipantLikes> allLikes,
        @JsonProperty("common_likes") List<MovieData> commonLikes
    ) implements ServerMessage {}

    record StreamingEnded() implements ServerMessage {}

    record MatchFound(
        @JsonProperty("all_likes") List<ParticipantLikes> allLikes,
        @JsonProperty("common_likes") List<MovieData> commonLikes
    ) implements ServerMessage {}

    record Error(String message) implements ServerMessage {}
}

File: movie-matcher/frontend-smartphone/lib/models/server_message.dart
import 'movie_data.dart';
import 'participant_likes.dart';

/// Серверные сообщения из WebSocket — соответствие ServerMessage из backend-java
/// Десериализация по полю "type"
sealed class ServerMessage {
  const ServerMessage();

  factory ServerMessage.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String;
    switch (type) {
      case 'ParticipantJoined':
        return ParticipantJoinedMessage(
          participantId: json['participant_id'] as String,
        );
      case 'ParticipantLeft':
        return ParticipantLeftMessage(
          participantId: json['participant_id'] as String,
        );
      case 'ParticipantReady':
        return ParticipantReadyMessage(
          participantId: json['participant_id'] as String,
          readyCount: json['ready_count'] as int,
          totalCount: json['total_count'] as int,
        );
      case 'VotingStarted':
        return const VotingStartedMessage();
      case 'MatchingStarted':
        return const MatchingStartedMessage();
      case 'RoomLocked':
        return RoomLockedMessage(message: json['message'] as String);
      case 'NewMovie':
        return NewMovieMessage(
          movie: MovieData.fromJson(json['movie'] as Map<String, dynamic>),
        );
      case 'VoteRecorded':
        return VoteRecordedMessage(
          participantId: json['participant_id'] as String,
          movieId: json['movie_id'] as String,
          isLike: json['is_like'] as bool,
        );
      case 'RoundCompleted':
        return RoundCompletedMessage(
          roundNumber: json['round_number'] as int,
          commonLikes: (json['common_likes'] as List<dynamic>)
              .map((e) => e as String)
              .toList(),
        );
      case 'VotingCompleted':
        return VotingCompletedMessage(
          matchedMovies: (json['matched_movies'] as List<dynamic>)
              .map((e) => e as String)
              .toList(),
        );
      case 'LikesUpdated':
        return LikesUpdatedMessage(
          allLikes: (json['all_likes'] as List<dynamic>)
              .map((e) =>
                  ParticipantLikes.fromJson(e as Map<String, dynamic>))
              .toList(),
          commonLikes: (json['common_likes'] as List<dynamic>)
              .map((e) => MovieData.fromJson(e as Map<String, dynamic>))
              .toList(),
        );
      case 'MatchingEnded':
        return MatchingEndedMessage(
          allLikes: (json['all_likes'] as List<dynamic>)
              .map((e) =>
                  ParticipantLikes.fromJson(e as Map<String, dynamic>))
              .toList(),
          commonLikes: (json['common_likes'] as List<dynamic>)
              .map((e) => MovieData.fromJson(e as Map<String, dynamic>))
              .toList(),
        );
      case 'MatchFound':
        return MatchFoundMessage(
          allLikes: (json['all_likes'] as List<dynamic>)
              .map((e) =>
                  ParticipantLikes.fromJson(e as Map<String, dynamic>))
              .toList(),
          commonLikes: (json['common_likes'] as List<dynamic>)
              .map((e) => MovieData.fromJson(e as Map<String, dynamic>))
              .toList(),
        );
      case 'StreamingEnded':
        return const StreamingEndedMessage();
      case 'NoMoreMovies':
        return const NoMoreMoviesMessage();
      case 'Error':
        return ErrorMessage(message: json['message'] as String);
      default:
        return ErrorMessage(message: 'Неизвестный тип сообщения: $type');
    }
  }
}

class ParticipantJoinedMessage extends ServerMessage {
  final String participantId;
  const ParticipantJoinedMessage({required this.participantId});
}

class ParticipantLeftMessage extends ServerMessage {
  final String participantId;
  const ParticipantLeftMessage({required this.participantId});
}

class ParticipantReadyMessage extends ServerMessage {
  final String participantId;
  final int readyCount;
  final int totalCount;
  const ParticipantReadyMessage({
    required this.participantId,
    required this.readyCount,
    required this.totalCount,
  });
}

class VotingStartedMessage extends ServerMessage {
  const VotingStartedMessage();
}

class MatchingStartedMessage extends ServerMessage {
  const MatchingStartedMessage();
}

class RoomLockedMessage extends ServerMessage {
  final String message;
  const RoomLockedMessage({required this.message});
}

class NewMovieMessage extends ServerMessage {
  final MovieData movie;
  const NewMovieMessage({required this.movie});
}

class VoteRecordedMessage extends ServerMessage {
  final String participantId;
  final String movieId;
  final bool isLike;
  const VoteRecordedMessage({
    required this.participantId,
    required this.movieId,
    required this.isLike,
  });
}

class RoundCompletedMessage extends ServerMessage {
  final int roundNumber;
  final List<String> commonLikes;
  const RoundCompletedMessage({
    required this.roundNumber,
    required this.commonLikes,
  });
}

class VotingCompletedMessage extends ServerMessage {
  final List<String> matchedMovies;
  const VotingCompletedMessage({required this.matchedMovies});
}

class LikesUpdatedMessage extends ServerMessage {
  final List<ParticipantLikes> allLikes;
  final List<MovieData> commonLikes;
  const LikesUpdatedMessage({
    required this.allLikes,
    required this.commonLikes,
  });
}

class MatchingEndedMessage extends ServerMessage {
  final List<ParticipantLikes> allLikes;
  final List<MovieData> commonLikes;
  const MatchingEndedMessage({
    required this.allLikes,
    required this.commonLikes,
  });
}

class MatchFoundMessage extends ServerMessage {
  final List<ParticipantLikes> allLikes;
  final List<MovieData> commonLikes;
  const MatchFoundMessage({
    required this.allLikes,
    required this.commonLikes,
  });
}

class StreamingEndedMessage extends ServerMessage {
  const StreamingEndedMessage();
}

class NoMoreMoviesMessage extends ServerMessage {
  const NoMoreMoviesMessage();
}

class ErrorMessage extends ServerMessage {
  final String message;
  const ErrorMessage({required this.message});
}
